src/query_builder.rs

```diff
@@ -0,0 +1,389 @@
+            [:db/add "s" :db/valueType :db.type/boolean]
+            [:db/add "s" :db/cardinality :db.cardinality/one]
+            [:db/add "t" :db/ident :foo/long]
+            [:db/add "t" :db/valueType :db.type/long]
+            [:db/add "t" :db/cardinality :db.cardinality/one]
+        ]"#).expect("successful transaction");
+
+        let report = store.transact(r#"[
+            [:db/add "l" :foo/boolean true]
+            [:db/add "l" :foo/long 25]
+            [:db/add "m" :foo/boolean false]
+            [:db/add "m" :foo/long 26]
+            [:db/add "n" :foo/boolean true]
+            [:db/add "n" :foo/long 27]
+            [:db/add "p" :foo/boolean false]
+            [:db/add "p" :foo/long 24]
+            [:db/add "u" :foo/boolean true]
+            [:db/add "u" :foo/long 23]
+        ]"#).expect("successful transaction");
+
+        let n_yes = report.tempids.get("n").expect("found it").clone();
+
+        let results: Vec<TypedValue> = QueryBuilder::new(&mut store, r#"[:find [?x, ?i]
+                                              :in ?v ?i
+                                              :where [?x :foo/boolean ?v]
+                                                     [?x :foo/long ?i]]"#)
+                              .bind_value("?v", true)
+                              .bind_long("?i", 27)
+                              .execute_tuple().expect("TupleResult").unwrap_or(vec![]);
+        let entid = TypedValue::Ref(n_yes.clone());
+        let long_val = TypedValue::Long(27);
+
+        assert_eq!(results, vec![entid, long_val]);
+    }
+
+    #[test]
+    fn test_rel_query_result() {
+        let mut store = Store::open("").expect("store connection");
+        store.transact(r#"[
+            [:db/add "s" :db/ident :foo/boolean]
+            [:db/add "s" :db/valueType :db.type/boolean]
+            [:db/add "s" :db/cardinality :db.cardinality/one]
+            [:db/add "t" :db/ident :foo/long]
+            [:db/add "t" :db/valueType :db.type/long]
+            [:db/add "t" :db/cardinality :db.cardinality/one]
+        ]"#).expect("successful transaction");
+
+        let report = store.transact(r#"[
+            [:db/add "l" :foo/boolean true]
+            [:db/add "l" :foo/long 25]
+            [:db/add "m" :foo/boolean false]
+            [:db/add "m" :foo/long 26]
+            [:db/add "n" :foo/boolean true]
+            [:db/add "n" :foo/long 27]
+        ]"#).expect("successful transaction");
+
+        let l_yes = report.tempids.get("l").expect("found it").clone();
+        let m_yes = report.tempids.get("m").expect("found it").clone();
+        let n_yes = report.tempids.get("n").expect("found it").clone();
+
+        #[derive(Debug, PartialEq)]
+        struct Res {
+            entid: i64,
+            boolean: bool,
+            long_val: i64,
+        };
+
+        let mut results: Vec<Res> = QueryBuilder::new(&mut store, r#"[:find ?x ?v ?i
+                                              :where [?x :foo/boolean ?v]
```

Comment:
Indenting for all of these test queries, plz.

Reply:
Fucking VSCode keeps realigning them, but only in tests. Grrr.